src/testRunner/unittests/tsserver/updateFileSystem.ts

```diff
@@ -0,0 +1,125 @@
+namespace ts.projectSystem {
+    describe("unittests:: tsserver:: updateFileSystem", () => {
+        const config: protocol.FileSystemRequestArgs = {
+            file: "/fshost/b/tsconfig.json",
+            fileContent: "{}"
+        };
+        const file3: protocol.FileSystemRequestArgs = {
+            file: "/fshost/b/file3.ts",
+            fileContent: "export let xyz = 1;"
+        };
+        const app: protocol.FileSystemRequestArgs = {
+            file: "/fshost/b/app.ts",
+            fileContent: "import { xyz } from './file3'; let x = xyz"
+        };
+        const file1: protocol.FileSystemRequestArgs = {
+            file: "/fshost/b/commonFile1.ts",
+            fileContent: "let x = 1"
+        };
+        const file2: protocol.FileSystemRequestArgs = {
+            file: "/fshost/b/commonFile2.ts",
+            fileContent: "let y = 1"
+        };
+        const lib: protocol.FileSystemRequestArgs = {
+        file: "/fshost/lib/lib.d.ts",
+        fileContent: `/// <reference no-default-lib="true"/>
+interface Boolean {}
+interface Function {}
+interface CallableFunction {}
+interface NewableFunction {}
+interface IArguments {}
+interface Number { toExponential: any; }
+interface Object {}
+interface RegExp {}
+interface String { charAt: any; }
+interface Array<T> { length: number; [n: number]: T; }`
+    };
+
+        function fileContentWithComment(file: protocol.FileSystemRequestArgs) {
+            return `// some copy right notice
+${file.fileContent}`;
+        }
+        function baselineFileSystem(scenario: string, subScenario: string, requests: [string, Partial<protocol.Request>][], host: VirtualFS.VirtualServerHost, session: TestSession) {

```

Comment:
Instead of passing host here.. it would be good to diff host and fsHost after each request. 
Also instead of taking request you can take `callback | request` so you can execute callback and do diff.. there you would execute the ensureFileOrFolder etc kind of tests